### PR TITLE
Add Infrastructure Libraries to NPSP

### PR DESCRIPTION
**Pull request by kenlewis-sfdo**
This PR brings the following libraries into the NPSP
project: Apex Mocks, Apex Common, Apex Extensions.

The libraries are "vendored," meaning the repo has
its own copy of each library.  When new features are
introduced into the original libraries, an internal
"release" process will be used update the NPSP repo
version of each library, on a schedule determined
by the NPSP engineering teams.

Apex Mocks: https://github.com/apex-enterprise-patterns/fflib-apex-mocks
Run-time mock configuration.  Reduce lines of code needed
to configure various test configurations.  Empower engineers
to write more variations of tests. Enable easier unit-testing.
Dependency of Apex Common.

Apex Common: https://github.com/apex-enterprise-patterns/fflib-apex-common
Base Apex Enterprise Patterns library.  Empowers engineers to
use established patterns and base class methods to support
separation of concerns, consistent techniques for interacting
with platform APIs, querying, security, DML, etc.

Apex Extensions: https://github.com/wimvelzeboer/fflib-apex-extensions
Extensions package for Apex Common.  Includes many features
beyond Apex Common, including many additional domain
methods, dependency injection via configuration, etc.

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Review by lparrott — APPROVED** [review]
Looks good to me! The only thing I noticed is that `FieldMappings` is still extending `fflib_SObjects` instead of `fflib_SObjects2` 🤷‍♂️

**Review by psadie-sfdo — APPROVED** [review]
Just two small naming notes!

1. Looks like there's small typo for the following directory `force-app/infrastructure/apex-extensions/main/default/classes/triggerHanders` -> `../triggerHandlers`.
2. Pluralize the child directories of `force-app/main` to align with naming in the `force-app/infrastructure`.

Otherwise looks good to me as well! I appreciated the resources and vids in the design doc!

**Comment by kenlewis-sfdo** [comment]
> Just two small naming notes!
> 
> 1. Looks like there's small typo for the following directory `force-app/infrastructure/apex-extensions/main/default/classes/triggerHanders` -> `../triggerHandlers`.
> 2. Pluralize the child directories of `force-app/main` to align with naming in the `force-app/infrastructure`.
> 
> Otherwise looks good to me as well! I appreciated the resources and vids in the design doc!

Thanks Prince! I updated the trigger handler folder name but left those in our main directory just to keep the diff smaller.  I don't have any issues with that change, just think it is minor and not necessary right now. Let me know if you feel differently. Thanks for the review!

**Comment by kenlewis-sfdo** [comment]
@force2b Just wanted to ping you that we are close on this one.  Please do let us know if you have any outstanding concerns.  Thank you!

**Review by bdvorachek — APPROVED** [review]
Overall this looks good. One concern I have here is that fflib_SecurityUtils here uses some new custom labels that we're introducing as part of this infrastructure library. We should take care that any exception thrown by that class does not make it up to the UI, as it would contain field/object info.

File: force-app/infrastructure/apex-extensions/main/default/classes/domains/fflib_SObjects2.cls [hunk 1]
```diff
@@ -0,0 +1,840 @@
...
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+public virtual class fflib_SObjects2
```
Comment (force2b) [inline force-app/infrastructure/apex-extensions/main/default/classes/domains/fflib_SObjects2.cls:31]: Why rename this to`fflib_Sobjects2`?

Comment (kenlewis-sfdo) [reply force-app/infrastructure/apex-extensions/main/default/classes/domains/fflib_SObjects2.cls:31]: Its not a rename.  SObjects2(extensions package) extends SObjects(which is in the common package) to  provide more base methods.

**Review by force2b — APPROVED** [review]

**Pull request merged**